File: Torch/JoinTable.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

--[[
Assume all the inputs have identical SparseGrids and input[i].features:size(1)
Assume input[1].features:size(2) >= input[i].features:size(2) for all i=1,#input
output.validRules is taken from input[1].validRules (could do set union?)
(for resnets, make sure the residual link is input[2])
]]

return function(sparseconvnet)
  local JoinTable, parent = torch.class(
    'sparseconvnet.JoinTable', 'nn.Module', sparseconvnet)

  function JoinTable:__init(nPlanes)
    parent.__init(self)
    self.nOutputPlanes=0
    self.nPlanes=nPlanes
    self.gradInput = {}
    for i,j in ipairs(nPlanes) do
      self.gradInput[i]={features=torch.Tensor()}
      self.nOutputPlanes=self.nOutputPlanes+j
    end
    self.output = {
      features = torch.Tensor()
    }
  end

  function JoinTable:updateOutput(input)
    self.output.features:resize(input[1].features:size(1),self.nOutputPlanes)
    self.output.metadata=input[1].metadata
    self.output.spatialSize=input[1].spatialSize
    local offset=0
    for i,j in ipairs(self.nPlanes) do
      assert(input[i].features:size(1)==input[1].features:size(1))
      assert(input[i].features:size(2)==j)
      self.output.features:narrow(2,1+offset,j):copy(input[i].features)
      offset=offset+j
    end
    return self.output
  end

  function JoinTable:updateGradInput(input, gradOutput)
    local offset=0
    for i,j in ipairs(self.nPlanes) do
      self.gradInput[i].features:resize(input[1].features:size(1),j):copy(gradOutput.features:narrow(2,1+offset,j))
      offset=offset+j
    end
    return self.gradInput
  end

  function JoinTable:clearState()
    self.output = {
      features = self.output.features:set()
    }
    for i,j in ipairs(self.nPlanes) do
      self.gradInput[i]={features=self.gradInput[i].features:set()}
    end
  end

  function JoinTable:suggestInputSize(nOut)
    return nOut
  end
end

File: PyTorch/sparseconvnet/legacy/joinTable.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch
import sparseconvnet
from . import SparseModule
from ..utils import toLongTensor, dim_typed_fn, optionalTensor, nullptr, set
from ..sparseConvNetTensor import SparseConvNetTensor


class JoinTable(SparseModule):
    def __init__(self, nPlanes):
        SparseModule.__init__(self)
        self.nPlanes = nPlanes
        self.gradInput = [torch.Tensor() for p in nPlanes]
        self.nOutputPlanes = sum(nPlanes)
        self.output = SparseConvNetTensor(torch.Tensor())

    def updateOutput(self, input):
        self.output.features.resize_(
            input[0].features.size(0),
            self.nOutputPlanes)
        self.output.metadata = input[0].metadata
        self.output.spatial_size = input[0].spatial_size
        offset = 0
        for i, n in zip(input, self.nPlanes):
            self.output.features.narrow(1, offset, n).copy_(i.features)
            offset += n
        return self.output

    def updateGradInput(self, input, gradOutput):
        offset = 0
        a = input[0].features.size(0)
        for g, n in zip(self.gradInput, self.nPlanes):
            g.resize_(a, n).copy_(gradOutput.narrow(1, offset, n))
            offset += n
        return self.gradInput

    def type(self, t, tensorCache=None):
        if t:
            self.output.type(t)
            self.gradInput = [g.type(t) for g in self.gradInput]

    def clearState(self):
        set(self.output)
        for g in self.gradInput:
            set(g)

    def __repr__(self):
        s = 'JoinTable: ' + str(self.nPlanes[0])
        for n in self.nPlanes[1:]:
            s = s + ' + ' + str(n)
        s = s + ' -> ' + str(self.nOutputPlanes)
        return s
